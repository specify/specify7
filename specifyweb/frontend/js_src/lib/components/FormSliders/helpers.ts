import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
