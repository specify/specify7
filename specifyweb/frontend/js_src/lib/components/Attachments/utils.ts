import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
