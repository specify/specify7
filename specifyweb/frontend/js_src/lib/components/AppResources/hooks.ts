import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useAsyncState } from '../../hooks/useAsyncState';
import { f } from '../../utils/functools';
import type { GetOrSet, IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import { fetchCollection } from '../DataModel/collection';
import type { SerializedResource } from '../DataModel/helperTypes';
import type {
  Collection,
  Discipline,
  SpAppResource,
  SpAppResourceData,
  SpecifyUser,
  SpViewSetObj,
} from '../DataModel/types';
import { userPreferences } from '../Preferences/userPreferences';
import { getAppResourceType } from './filtersHelpers';
import { getAppResourceCount, getAppResourceMode } from './helpers';
import { getAppResourceTree, getScope } from './tree';
import type { ScopedAppResourceDir } from './types';
import { appResourceSubTypes } from './types';

export type AppResources = {
  readonly directories: RA<ScopedAppResourceDir>;
  readonly disciplines: RA<SerializedResource<Discipline>>;
  readonly collections: RA<SerializedResource<Collection>>;
  readonly users: RA<SerializedResource<SpecifyUser>>;
  readonly appResources: RA<SerializedResource<SpAppResource>>;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
};

export function useAppResources(
  loadingScreen: boolean = true
): GetOrSet<AppResources | undefined> {
  return useAsyncState(
    React.useCallback(
      async () =>
        f.all({
          directories: fetchCollection('SpAppResourceDir', {
            limit: 0,
            domainFilter: false,
          }).then<AppResources['directories']>(({ records }) =>
            records.map((record) => ({ ...record, scope: getScope(record) }))
          ),
          disciplines: fetchCollection('Discipline', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          collections: fetchCollection('Collection', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          users: fetchCollection('SpecifyUser', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          appResources: fetchCollection('SpAppResource', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
          viewSets: fetchCollection('SpViewSetObj', {
            limit: 0,
            domainFilter: false,
          }).then(({ records }) => records),
        }),
      []
    ),
    loadingScreen
  );
}

export type AppResourcesTree = RA<{
  readonly label: LocalizedString;
  /*
   * A string that would be stable thought the lifecycle of an object.
   * Used to identify a tree node when storing conformation it in localStorage.
   *
   * Directory key is created instead of using id, because app resource
   * needs a way to identify directories that don't yet have IDs (i.e, when
   * creating an app resource for a collection that didn't have any app resources
   * before, the collection won't have any SpAppResourceDir record, thus not
   * directory ID).
   */
  readonly key: string;
  readonly directory: ScopedAppResourceDir | undefined;
  readonly appResources: RA<
    SerializedResource<SpAppResource> & {
      readonly label?: LocalizedString;
    }
  >;
  readonly viewSets: RA<SerializedResource<SpViewSetObj>>;
  readonly subCategories: AppResourcesTree;
}>;

export function useResourcesTree(resources: AppResources): AppResourcesTree {
  const [localize] = userPreferences.use(
    'appResources',
    'appearance',
    'localizeResourceNames'
  );
  return React.useMemo<AppResourcesTree>(() => {
    const tree = getAppResourceTree(resources);
    return localize ? localizeTree(tree) : tree;
  }, [resources, localize]);
}

const localizeTree = (tree: AppResourcesTree): AppResourcesTree =>
  tree.map(({ appResources, subCategories, ...rest }) => ({
    ...rest,
    appResources: appResources.map(localizeResource),
    subCategories: localizeTree(subCategories),
  }));

function localizeResource(
  resource: SerializedResource<SpAppResource> & {
    readonly label?: LocalizedString;
  }
): SerializedResource<SpAppResource> & { readonly label?: LocalizedString } {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  // Check that resource of this type can only have one specific name
  return typeof type.name === 'string'
    ? {
        ...resource,
        // Then replace the name with a localized label unless it's already set
        label: resource.label ?? type.label,
      }
    : resource;
}

export function useAppResourceCount(
  resourcesTree: AppResourcesTree[number]
): number {
  return React.useMemo(
    () => getAppResourceCount(resourcesTree),
    [resourcesTree]
  );
}

/**
 * Fetch resource contents
 */
export function useAppResourceData(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  initialData: string | undefined
): {
  readonly resourceData: GetOrSet<
    SerializedResource<SpAppResourceData> | undefined
  >;
  readonly isChanged: boolean;
} {
  const initialValue = React.useRef<string>('');
  const [resourceData, setResourceData] = useAsyncState<
    SerializedResource<SpAppResourceData>
  >(
    React.useCallback(async () => {
      const relationshipName =
        getAppResourceMode(resource) === 'appResources'
          ? 'spAppResource'
          : 'spViewSetObj';
      const newResource = addMissingFields('SpAppResourceData', {
        [relationshipName]: resource.id,
        data: initialData,
      });
      const dataResource =
        typeof resource.id === 'number'
          ? await fetchCollection('SpAppResourceData', {
              limit: 1,
              [relationshipName]: resource.id,
              domainFilter: false,
            }).then(
              ({ records }) =>
                /*
                 * For some reason, app resource can have multiple app resource
                 * datas (but it never does in practice)
                 */
                records[0] ?? newResource
            )
          : newResource;
      const newData = fixLineBreaks(dataResource.data ?? '');
      initialValue.current = newData;
      return { ...dataResource, data: newData };
    }, [resource, initialData]),
    true
  );
  return {
    resourceData: [resourceData, setResourceData],
    isChanged: initialValue.current !== resourceData?.data,
  };
}

const fixLineBreaks = (string: string): string =>
  string.replaceAll(/[\n\r]+/gu, '\n');

export const getAppResourceExtension = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): string =>
  resource._tableName === 'SpViewSetObj'
    ? 'xml'
    : getResourceExtension(resource as SerializedResource<SpAppResource>);

function getResourceExtension(
  resource: SerializedResource<SpAppResource>
): 'jrxml' | 'json' | 'properties' | 'txt' | 'xml' {
  const type = appResourceSubTypes[getAppResourceType(resource)];
  const mimeType = resource.mimeType?.toLowerCase() ?? type?.mimeType ?? '';
  if (mimeType in mimeMapper) return mimeMapper[mimeType];
  else if (mimeType.startsWith('jrxml')) return 'jrxml';
  else if (
    resource.name === 'preferences' &&
    mimeType === appResourceSubTypes.otherPropertiesResource.mimeType
  )
    return 'properties';
  else return 'txt';
}

const mimeMapper: IR<'json' | 'properties' | 'txt' | 'xml'> = {
  'application/json': 'json',
  'text/xml': 'xml',
  'text/plain': 'txt',
};
