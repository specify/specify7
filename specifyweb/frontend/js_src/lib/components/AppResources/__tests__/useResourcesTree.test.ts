import { renderHook } from '@testing-library/react';

import { requireContext } from '../../../tests/helpers';
import { getAppResourceCount } from '../helpers';
import type { AppResourcesTree } from '../hooks';
import { useResourcesTree } from '../hooks';
import { staticAppResources } from './staticAppResources';
import { utilsForTests } from './utils';

requireContext();

const { setAppResourceDir, testDisciplines } = utilsForTests;

describe('useResourcesTree', () => {
  const getResourceCountTree = (result: AppResourcesTree) =>
    result.reduce(
      (count, resource) => count + getAppResourceCount(resource),
      0
    );

  /*
   * Const getResourceCountStatic = (resources: AppResources) => {
   *     const allAppResources = [
   *         ...resources.appResources.map(({resource_uri})=>resource_uri),
   *         ...resources.viewSets.map(({resource_uri})=>resource_uri)
   *     ]
   *     return allAppResources.length;
   * }
   */

  const resources = {
    ...staticAppResources,
    disciplines: testDisciplines,
  };

  test('missing appresource dir', () => {
    const { result } = renderHook(() => useResourcesTree(resources));

    expect(result.current).toMatchSnapshot();

    // There is only 1 resource with the matching spappresourcedir.
    expect(getResourceCountTree(result.current)).toBe(1);
  });

  test('all appresource dir', () => {
    const appResourceSet = resources.appResources.reduce(
      (resources, _, index) =>
        setAppResourceDir(resources, 'appResources', index, 4),
      resources
    );

    const viewSet = setAppResourceDir(appResourceSet, 'viewSets', 0, 4);

    const { result } = renderHook(() => useResourcesTree(viewSet));

    expect(result.current).toMatchSnapshot();

    expect(getResourceCountTree(result.current)).toBe(4);
  });
});
