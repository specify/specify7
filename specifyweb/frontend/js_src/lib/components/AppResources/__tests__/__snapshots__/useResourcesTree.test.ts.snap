// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResourcesTree all appresource dir 1`] = `
[
  {
    "appResources": [],
    "directory": {
      "_tableName": "SpAppResourceDir",
      "collection": null,
      "createdByAgent": "/api/specify/agent/3/",
      "discipline": null,
      "disciplineType": null,
      "id": 3,
      "isPersonal": false,
      "modifiedByAgent": null,
      "resource_uri": "/api/specify/spappresourcedir/3/",
      "scope": "global",
      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=3",
      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=3",
      "specifyUser": null,
      "timestampCreated": "2012-08-10T16:12:08",
      "timestampModified": "2012-08-10T16:12:08",
      "userType": "Global Prefs",
      "version": 683,
    },
    "key": "globalResource",
    "label": "Global Resources",
    "subCategories": [],
    "viewSets": [],
  },
  {
    "appResources": [],
    "directory": undefined,
    "key": "disciplineResources",
    "label": "Discipline Resources",
    "subCategories": [
      {
        "appResources": [],
        "directory": {
          "_tableName": "SpAppResourceDir",
          "collection": "/api/specify/collection/4/",
          "createdByAgent": null,
          "discipline": "/api/specify/discipline/3/",
          "disciplineType": null,
          "isPersonal": false,
          "modifiedByAgent": null,
          "resource_uri": undefined,
          "specifyUser": null,
          "timestampCreated": "2022-08-31",
          "timestampModified": null,
          "userType": null,
          "version": 1,
        },
        "key": "discipline_3",
        "label": "Ichthyology",
        "subCategories": [
          {
            "appResources": [],
            "directory": {
              "_tableName": "SpAppResourceDir",
              "collection": "/api/specify/collection/65536/",
              "createdByAgent": null,
              "discipline": "/api/specify/discipline/3/",
              "disciplineType": null,
              "isPersonal": false,
              "modifiedByAgent": null,
              "resource_uri": undefined,
              "specifyUser": null,
              "timestampCreated": "2022-08-31",
              "timestampModified": null,
              "userType": null,
              "version": 1,
            },
            "key": "collection_65536",
            "label": "KUFishTeaching",
            "subCategories": [
              {
                "appResources": [],
                "directory": undefined,
                "key": "users",
                "label": "User Accounts",
                "subCategories": [
                  {
                    "appResources": [
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": "QueryExtraList",
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 3,
                        "level": 0,
                        "metaData": null,
                        "mimeType": "text/xml",
                        "modifiedByAgent": "/api/specify/agent/3/",
                        "name": "QueryExtraList",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/3/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=3",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=3",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2012-08-10T16:12:05",
                        "version": 2,
                      },
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": null,
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 4,
                        "level": 3,
                        "metaData": null,
                        "mimeType": null,
                        "modifiedByAgent": null,
                        "name": "preferences",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/4/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=4",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=4",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2012-08-10T16:12:08",
                        "version": 683,
                      },
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": null,
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 73,
                        "label": "Default User Preferences",
                        "level": 0,
                        "metaData": null,
                        "mimeType": "application/json",
                        "modifiedByAgent": null,
                        "name": "DefaultUserPreferences",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/73/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=73",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=73",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2025-07-04T00:00:00",
                        "version": 1,
                      },
                    ],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": "/api/specify/agent/3/",
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": "Ichthyology",
                      "id": 4,
                      "isPersonal": true,
                      "modifiedByAgent": null,
                      "resource_uri": "/api/specify/spappresourcedir/4/",
                      "scope": "user",
                      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=4",
                      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=4",
                      "specifyUser": "/api/specify/specifyuser/4/",
                      "timestampCreated": "2012-10-01T10:29:36",
                      "timestampModified": "2012-10-01T10:29:36",
                      "userType": "manager",
                      "version": 3,
                    },
                    "key": "collection_65536_user_5",
                    "label": "cmeyer",
                    "subCategories": [],
                    "viewSets": [
                      {
                        "_tableName": "SpViewSetObj",
                        "createdByAgent": "/api/specify/agent/2151/",
                        "description": null,
                        "fileName": null,
                        "id": 5,
                        "level": 2,
                        "metaData": null,
                        "modifiedByAgent": null,
                        "name": "fish.views",
                        "resource_uri": "/api/specify/spviewsetobj/5/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spviewsetobj=5",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "timestampCreated": "2014-01-28T08:44:29",
                        "timestampModified": "2014-01-28T08:44:29",
                        "version": 6,
                      },
                    ],
                  },
                  {
                    "appResources": [
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": "QueryExtraList",
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 3,
                        "level": 0,
                        "metaData": null,
                        "mimeType": "text/xml",
                        "modifiedByAgent": "/api/specify/agent/3/",
                        "name": "QueryExtraList",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/3/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=3",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=3",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2012-08-10T16:12:05",
                        "version": 2,
                      },
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": null,
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 4,
                        "level": 3,
                        "metaData": null,
                        "mimeType": null,
                        "modifiedByAgent": null,
                        "name": "preferences",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/4/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=4",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=4",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2012-08-10T16:12:08",
                        "version": 683,
                      },
                      {
                        "_tableName": "SpAppResource",
                        "allPermissionLevel": null,
                        "createdByAgent": "/api/specify/agent/3/",
                        "description": null,
                        "group": null,
                        "groupPermissionLevel": null,
                        "id": 73,
                        "label": "Default User Preferences",
                        "level": 0,
                        "metaData": null,
                        "mimeType": "application/json",
                        "modifiedByAgent": null,
                        "name": "DefaultUserPreferences",
                        "ownerPermissionLevel": null,
                        "resource_uri": "/api/specify/spappresource/73/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=73",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "spReports": "/api/specify/spreport/?appresource=73",
                        "specifyUser": "/api/specify/specifyuser/4/",
                        "timestampCreated": "2025-07-04T00:00:00",
                        "version": 1,
                      },
                    ],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": "/api/specify/agent/3/",
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": "Ichthyology",
                      "id": 4,
                      "isPersonal": true,
                      "modifiedByAgent": null,
                      "resource_uri": "/api/specify/spappresourcedir/4/",
                      "scope": "user",
                      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=4",
                      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=4",
                      "specifyUser": "/api/specify/specifyuser/4/",
                      "timestampCreated": "2012-10-01T10:29:36",
                      "timestampModified": "2012-10-01T10:29:36",
                      "userType": "manager",
                      "version": 3,
                    },
                    "key": "collection_65536_user_4",
                    "label": "Vertnet",
                    "subCategories": [],
                    "viewSets": [
                      {
                        "_tableName": "SpViewSetObj",
                        "createdByAgent": "/api/specify/agent/2151/",
                        "description": null,
                        "fileName": null,
                        "id": 5,
                        "level": 2,
                        "metaData": null,
                        "modifiedByAgent": null,
                        "name": "fish.views",
                        "resource_uri": "/api/specify/spviewsetobj/5/",
                        "spAppResourceDatas": "/api/specify/spappresourcedata/?spviewsetobj=5",
                        "spAppResourceDir": "/api/specify/spappresourcedir/4/",
                        "timestampCreated": "2014-01-28T08:44:29",
                        "timestampModified": "2014-01-28T08:44:29",
                        "version": 6,
                      },
                    ],
                  },
                ],
                "viewSets": [],
              },
              {
                "appResources": [],
                "directory": undefined,
                "key": "userTypes",
                "label": "User Types",
                "subCategories": [
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "fullaccess",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_FullAccess",
                    "label": "FullAccess",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "guest",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_Guest",
                    "label": "Guest",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "limitedaccess",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_LimitedAccess",
                    "label": "LimitedAccess",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "manager",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_Manager",
                    "label": "Manager",
                    "subCategories": [],
                    "viewSets": [],
                  },
                ],
                "viewSets": [],
              },
            ],
            "viewSets": [],
          },
        ],
        "viewSets": [],
      },
    ],
    "viewSets": [],
  },
]
`;

exports[`useResourcesTree missing appresource dir 1`] = `
[
  {
    "appResources": [
      {
        "_tableName": "SpAppResource",
        "allPermissionLevel": null,
        "createdByAgent": "/api/specify/agent/3/",
        "description": null,
        "group": null,
        "groupPermissionLevel": null,
        "id": 4,
        "label": "Global Preferences",
        "level": 3,
        "metaData": null,
        "mimeType": null,
        "modifiedByAgent": null,
        "name": "preferences",
        "ownerPermissionLevel": null,
        "resource_uri": "/api/specify/spappresource/4/",
        "spAppResourceDatas": "/api/specify/spappresourcedata/?spappresource=4",
        "spAppResourceDir": "/api/specify/spappresourcedir/3/",
        "spReports": "/api/specify/spreport/?appresource=4",
        "specifyUser": "/api/specify/specifyuser/4/",
        "timestampCreated": "2012-08-10T16:12:08",
        "version": 683,
      },
    ],
    "directory": {
      "_tableName": "SpAppResourceDir",
      "collection": null,
      "createdByAgent": "/api/specify/agent/3/",
      "discipline": null,
      "disciplineType": null,
      "id": 3,
      "isPersonal": false,
      "modifiedByAgent": null,
      "resource_uri": "/api/specify/spappresourcedir/3/",
      "scope": "global",
      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=3",
      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=3",
      "specifyUser": null,
      "timestampCreated": "2012-08-10T16:12:08",
      "timestampModified": "2012-08-10T16:12:08",
      "userType": "Global Prefs",
      "version": 683,
    },
    "key": "globalResource",
    "label": "Global Resources",
    "subCategories": [],
    "viewSets": [],
  },
  {
    "appResources": [],
    "directory": undefined,
    "key": "disciplineResources",
    "label": "Discipline Resources",
    "subCategories": [
      {
        "appResources": [],
        "directory": {
          "_tableName": "SpAppResourceDir",
          "collection": "/api/specify/collection/4/",
          "createdByAgent": null,
          "discipline": "/api/specify/discipline/3/",
          "disciplineType": null,
          "isPersonal": false,
          "modifiedByAgent": null,
          "resource_uri": undefined,
          "specifyUser": null,
          "timestampCreated": "2022-08-31",
          "timestampModified": null,
          "userType": null,
          "version": 1,
        },
        "key": "discipline_3",
        "label": "Ichthyology",
        "subCategories": [
          {
            "appResources": [],
            "directory": {
              "_tableName": "SpAppResourceDir",
              "collection": "/api/specify/collection/65536/",
              "createdByAgent": null,
              "discipline": "/api/specify/discipline/3/",
              "disciplineType": null,
              "isPersonal": false,
              "modifiedByAgent": null,
              "resource_uri": undefined,
              "specifyUser": null,
              "timestampCreated": "2022-08-31",
              "timestampModified": null,
              "userType": null,
              "version": 1,
            },
            "key": "collection_65536",
            "label": "KUFishTeaching",
            "subCategories": [
              {
                "appResources": [],
                "directory": undefined,
                "key": "users",
                "label": "User Accounts",
                "subCategories": [
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": "/api/specify/agent/3/",
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": "Ichthyology",
                      "id": 4,
                      "isPersonal": true,
                      "modifiedByAgent": null,
                      "resource_uri": "/api/specify/spappresourcedir/4/",
                      "scope": "user",
                      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=4",
                      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=4",
                      "specifyUser": "/api/specify/specifyuser/4/",
                      "timestampCreated": "2012-10-01T10:29:36",
                      "timestampModified": "2012-10-01T10:29:36",
                      "userType": "manager",
                      "version": 3,
                    },
                    "key": "collection_65536_user_5",
                    "label": "cmeyer",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": "/api/specify/agent/3/",
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": "Ichthyology",
                      "id": 4,
                      "isPersonal": true,
                      "modifiedByAgent": null,
                      "resource_uri": "/api/specify/spappresourcedir/4/",
                      "scope": "user",
                      "spPersistedAppResources": "/api/specify/spappresource/?spappresourcedir=4",
                      "spPersistedViewSets": "/api/specify/spviewsetobj/?spappresourcedir=4",
                      "specifyUser": "/api/specify/specifyuser/4/",
                      "timestampCreated": "2012-10-01T10:29:36",
                      "timestampModified": "2012-10-01T10:29:36",
                      "userType": "manager",
                      "version": 3,
                    },
                    "key": "collection_65536_user_4",
                    "label": "Vertnet",
                    "subCategories": [],
                    "viewSets": [],
                  },
                ],
                "viewSets": [],
              },
              {
                "appResources": [],
                "directory": undefined,
                "key": "userTypes",
                "label": "User Types",
                "subCategories": [
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "fullaccess",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_FullAccess",
                    "label": "FullAccess",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "guest",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_Guest",
                    "label": "Guest",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "limitedaccess",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_LimitedAccess",
                    "label": "LimitedAccess",
                    "subCategories": [],
                    "viewSets": [],
                  },
                  {
                    "appResources": [],
                    "directory": {
                      "_tableName": "SpAppResourceDir",
                      "collection": "/api/specify/collection/65536/",
                      "createdByAgent": null,
                      "discipline": "/api/specify/discipline/3/",
                      "disciplineType": null,
                      "isPersonal": false,
                      "modifiedByAgent": null,
                      "resource_uri": undefined,
                      "specifyUser": null,
                      "timestampCreated": "2022-08-31",
                      "timestampModified": null,
                      "userType": "manager",
                      "version": 1,
                    },
                    "key": "collection_65536_userType_Manager",
                    "label": "Manager",
                    "subCategories": [],
                    "viewSets": [],
                  },
                ],
                "viewSets": [],
              },
            ],
            "viewSets": [],
          },
        ],
        "viewSets": [],
      },
    ],
    "viewSets": [],
  },
]
`;
